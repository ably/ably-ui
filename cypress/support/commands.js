Cypress.Commands.add(
  "compareComponents",
  ({
    reactComponentUrl,
    viewComponentUrl,
    beforeScreenshot = () => {},
    selector,
    componentName = Cypress.mocha.getRunner().suite.ctx.test.parent.title,
  }) => {
    let reactScreenshotPath = null;
    let vwScreenshotPath = null;

    cy.exec(`mkdir -p cypress/screenshots/component-parity/results`, {
      log: false,
    });

    const diffPath = `cypress/screenshots/component-parity/results/${componentName}.png`;

    // If you are wondering why this isn't Promise.all or async/await have a look at
    // https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

    cy.visit(reactComponentUrl);

    beforeScreenshot();

    cy.get(selector)
      .screenshot(`${componentName}.react`, {
        onAfterScreenshot: (_, props) => (reactScreenshotPath = props.path),
      })
      .then(() => {
        cy.visit(viewComponentUrl);
        beforeScreenshot();

        cy.get(selector)
          .screenshot(`${componentName}.vw`, {
            onAfterScreenshot: (_, props) => (vwScreenshotPath = props.path),
          })
          .then(() => {
            cy.exec(
              `cypress/support/diff.js ` +
                `--pathA="${reactScreenshotPath}" ` +
                `--pathB="${vwScreenshotPath}" ` +
                `--target="${diffPath}" ` +
                `--threshold="0.005" `
            ).then((result) => {
              cy.exec(`rm "${reactScreenshotPath}" "${vwScreenshotPath}"`, {
                log: false,
              });

              assert.isTrue(result.stdout === "Yay", "Screenshots match");
            });
          });
      });
  }
);
