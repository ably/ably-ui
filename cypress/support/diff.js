#! /usr/bin/env node

// Used by the compareComponents command

const Blink = require("blink-diff");
const argv = require("yargs").argv;

const diff = new Blink({
  imageAPath: argv.pathA,
  imageBPath: argv.pathB,
  imageOutputPath: argv.target,
  threshold: 0.05,
});

diff.run((error, result) => {
  // We read stdout/stderr - this script is executed interactively by cypress
  if (error) {
    console.error(error);
  } else {
    console.log(diff.hasPassed(result.code) ? "Yay" : "Nay");
  }
});
