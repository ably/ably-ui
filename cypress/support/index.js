// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import "cypress-plugin-snapshots/commands";
import "./commands";

// Cypress has a bug that breaks snapshot tests when running the whole suite
// Follow https://github.com/cypress-io/cypress/issues/3090
// Recommended workaround below:
export const fixSnapshotSpec = (filename) => () => {
  const path = require("path");
  // const relative = filename.substr(1); // removes leading "/"
  const projectRoot = Cypress.config("projectRoot");
  const absolute = path.join(projectRoot, filename);

  Cypress.spec = {
    absolute,
    name: path.basename(filename),
    filename,
  };
};
