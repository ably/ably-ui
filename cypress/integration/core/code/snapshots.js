
import { fixSnapshotSpec } from "../../../support";

beforeEach(fixSnapshotSpec(__filename));

describe("Code Snapshot Test", () => {
  const url = "/components/code";
  const sharedSpecs = (url) => {
    beforeEach(() => {
      cy.visit(url);
    });

    it("test case common to both", () => {
      cy.get("[data-id='code']").toMatchSnapshot();
    });
  };

  beforeEach(() => {
    cy.viewport("macbook-16");
  });

  describe("react", () => {
    sharedSpecs(url);
  });

  describe("view component", () => {
    sharedSpecs(`${url}?iframework=vw`);
  });
});