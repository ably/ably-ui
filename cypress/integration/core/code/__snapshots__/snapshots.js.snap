exports[`Code Snapshot Test > react > test case common to both #0`] = `
<div class="hljs overflow-auto p-32 " data-id="code">
  <pre
    lang="javascript"
  ><code class="language-javascript ui-text-code1"><span class="hljs-keyword">var</span> ably = <span class="hljs-keyword">new</span> Ably.Realtime(<span class="hljs-string">'1WChTA.mc0Biw:kNfiYG4KiPgmHHgH'</span>);
<span class="hljs-keyword">var</span> channel = ably.channels.get(<span class="hljs-string">'web-pal'</span>);

<span class="hljs-comment">// Subscribe to messages on channel</span>
channel.subscribe(<span class="hljs-string">'greeting'</span>, <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">message</span>) </span>{
  alert(message.data);
});</code></pre>
</div>
`;

exports[`Code Snapshot Test > react > test case common to both #1`] = `
<div class="hljs overflow-auto p-32 " data-id="code">
  <pre
    lang="swift"
  ><code class="language-swift ui-text-code1"><span class="hljs-keyword">let</span> ably <span class="hljs-operator">=</span> <span class="hljs-type">ARTRealtime</span>(key: <span class="hljs-string">"1WChTA.mc0Biw:kNfiYG4KiPgmHHgH"</span>)
<span class="hljs-keyword">let</span> channel <span class="hljs-operator">=</span> ably.channels.get(<span class="hljs-string">"web-pal"</span>)

<span class="hljs-comment">// Subscribe to messages on channel</span>
channel.subscribe(<span class="hljs-string">"greeting"</span>) { message <span class="hljs-keyword">in</span>
  <span class="hljs-built_in">print</span>(<span class="hljs-string">"<span class="hljs-subst">\\(message.data)</span>"</span>)
}</code></pre>
</div>
`;

exports[`Code Snapshot Test > react > test case common to both #2`] = `
<div class="hljs overflow-auto p-32 " data-id="code">
  <pre
    lang="java"
  ><code class="language-java ui-text-code1">AblyRealtime ably = <span class="hljs-keyword">new</span> AblyRealtime(<span class="hljs-string">"1WChTA.mc0Biw:kNfiYG4KiPgmHHgH"</span>);
Channel channel = ably.channels.get(<span class="hljs-string">"web-pal"</span>);

<span class="hljs-comment">/* Subscribe to messages on channel */</span>

MessageListener listener;
listener = <span class="hljs-keyword">new</span> MessageListener() {
  <span class="hljs-meta">@Override</span>
  <span class="hljs-function"><span class="hljs-keyword">public</span> <span class="hljs-keyword">void</span> <span class="hljs-title">onMessage</span><span class="hljs-params">(Message message)</span> </span>{
    System.out.print(message.data);
  };
};
channel.subscribe(<span class="hljs-string">"greeting"</span>, listener);</code></pre>
</div>
`;

exports[`Code Snapshot Test > view component > test case common to both #0`] = `
<div class="hljs overflow-auto p-32 " data-id="code">
  <pre
    lang="javascript"
  ><code class="language-javascript ui-text-code1"><span class="hljs-keyword">var</span> ably = <span class="hljs-keyword">new</span> Ably.Realtime(<span class="hljs-string">'1WChTA.mc0Biw:kNfiYG4KiPgmHHgH'</span>);
<span class="hljs-keyword">var</span> channel = ably.channels.get(<span class="hljs-string">'web-pal'</span>);

<span class="hljs-comment">// Subscribe to messages on channel</span>
channel.subscribe(<span class="hljs-string">'greeting'</span>, <span class="hljs-function"><span class="hljs-keyword">function</span>(<span class="hljs-params">message</span>) </span>{
  alert(message.data);
});</code></pre>
</div>
`;

exports[`Code Snapshot Test > view component > test case common to both #1`] = `
<div class="hljs overflow-auto p-32 " data-id="code">
  <pre
    lang="swift"
  ><code class="language-swift ui-text-code1"><span class="hljs-keyword">let</span> ably <span class="hljs-operator">=</span> <span class="hljs-type">ARTRealtime</span>(key: <span class="hljs-string">"1WChTA.mc0Biw:kNfiYG4KiPgmHHgH"</span>)
<span class="hljs-keyword">let</span> channel <span class="hljs-operator">=</span> ably.channels.get(<span class="hljs-string">"web-pal"</span>)

<span class="hljs-comment">// Subscribe to messages on channel</span>
channel.subscribe(<span class="hljs-string">"greeting"</span>) { message <span class="hljs-keyword">in</span>
  <span class="hljs-built_in">print</span>(<span class="hljs-string">"<span class="hljs-subst">\\(message.data)</span>"</span>)
}</code></pre>
</div>
`;

exports[`Code Snapshot Test > view component > test case common to both #2`] = `
<div class="hljs overflow-auto p-32 " data-id="code">
  <pre
    lang="java"
  ><code class="language-java ui-text-code1">AblyRealtime ably = <span class="hljs-keyword">new</span> AblyRealtime(<span class="hljs-string">"1WChTA.mc0Biw:kNfiYG4KiPgmHHgH"</span>);
Channel channel = ably.channels.get(<span class="hljs-string">"web-pal"</span>);

<span class="hljs-comment">/* Subscribe to messages on channel */</span>

MessageListener listener;
listener = <span class="hljs-keyword">new</span> MessageListener() {
  <span class="hljs-meta">@Override</span>
  <span class="hljs-function"><span class="hljs-keyword">public</span> <span class="hljs-keyword">void</span> <span class="hljs-title">onMessage</span><span class="hljs-params">(Message message)</span> </span>{
    System.out.print(message.data);
  };
};
channel.subscribe(<span class="hljs-string">"greeting"</span>, listener);</code></pre>
</div>
`;
